rfcs/0002_amethyst_as_bin.md

```diff
@@ -0,0 +1,144 @@
+# Table of Contents
+
+- [Motivation](#motivation)
+- [Guide Level Explanation](#guide-level-explanation)
+- [Reference Level Explanation](#reference-level-explanation)
+- [Drawbacks]
+- [Rationale and Alternatives](#rationale-and-alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+
+# Basic Info
+[basic]: #basic-info
+
+- Feature Name: amethyst_as_bin
+- Start Date: 2018-11-14
+- RFC PR: https://github.com/amethyst/rfcs/pull/6
+- [Tracking Issue](#tracking-issue): (leave this empty)
+
+After filling out these fields, please proceed to fill out the remaining sections of this document.
+
+# Summary
+[summary]: #summary
+
+This RFC is to act as a discussion of the merits and drawbacks of reversing our current relationship with our users.
+Currently they create the executable binary, and we create a crate for that binary to use.  However, if we instead
+chose to reverse that, and make it so that we provided a binary for which their crate hooked into we can now control
+`fn main()`.  After making this transition we would have to be much more dependent on our amethyst toolset than we currently are.
+
+# Motivation
+[motivation]: #motivation
+There are four areas of the software that stand to benefit from this
+
+1. Initialization. If we control the main engine initialization sequence then we can avoid handing control over to the user
+before the engine is fully initialized.  The biggest benefit this has is that user provided `System::setup()` methods can use
+`!Default` resources.  Most commonly, these resources come from the rendering system, such as `ScreenDimensions`.  In order to
+gain the most benefits from this we are going to have to introduce a "core dispatcher" concept to the engine.
+The core dispatcher would be responsible for input, and rendering, both of which would be optional.  If neither are enabled
+then we simply don't run the core dispatcher.
```

Comment:
I'd assume this would reduce parallelism since dispatchers are run sequentially.

Reply:
Maybe a bit, but I provide the following counterarguments

1. Rendering wasn't parallel anyways, it's a thread local system.

2. Input is pretty lightweight, mostly O(1).  It doesn't get more complicated the more entities there are for example.